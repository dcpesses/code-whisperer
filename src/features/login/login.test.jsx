/* eslint-env jest */
import {render, screen, fireEvent} from '@testing-library/react';
import {vi} from 'vitest';
import Login, {getLoginUrl} from './index';
import React from 'react';

vi.mock('../../../package.json', () => {
  return {
    ...vi.importActual('../../../package.json'),
    version: '0.0.0'
  };
});

describe('getLoginUrl', () => {
  test('should return twitch oauth url', () => {
    vi.stubEnv('VITE_APP_TWITCH_CLIENT_ID', 'VITE_APP_TWITCH_CLIENT_ID');
    vi.stubEnv('VITE_APP_TWITCH_CLIENT_SECRET', 'VITE_APP_TWITCH_CLIENT_SECRET');
    vi.stubEnv('VITE_APP_REDIRECT_URI', 'VITE_APP_REDIRECT_URI');

    expect(getLoginUrl()).toEqual('https://id.twitch.tv/oauth2/authorize?client_id=VITE_APP_TWITCH_CLIENT_ID&response_type=code&scope=chat:read chat:edit moderation:read user:manage:whispers moderator:read:chatters user:read:moderated_channels channel:read:vips channel:read:editors moderator:manage:announcements user:read:subscriptions&redirect_uri=VITE_APP_REDIRECT_URI');
  });
});

describe('Login', () => {
  describe('clearLocalStorageData', () => {
    test('should call setState', () => {
      vi.useFakeTimers({ toFake: ['nextTick'] });
      const component = new Login();
      vi.spyOn(component, 'setState').mockImplementation(()=>{});
      vi.spyOn(global, 'setTimeout');
      component.clearLocalStorageData();
      vi.advanceTimersByTime(5500);
      expect(global.setTimeout).toHaveBeenCalledTimes(1);
      global.setTimeout.mock.calls[0][0]();
      // console.log({calls: global.setTimeout.mock.calls[0][0]})

      expect(component.setState).toHaveBeenCalledTimes(2);
      vi.useRealTimers();
    });
  });
  describe('toggleChangelogModal', () => {
    test('should toggle the state of showChangelogModal', () => {
      const component = new Login();
      vi.spyOn(component, 'setState');
      component.state.showChangelogModal = false;
      component.toggleChangelogModal();
      expect(component.setState.mock.calls[0][0](component.state)).toEqual({ showChangelogModal: true });
    });
  });
  describe('render', () => {
    beforeEach(()=>{
      vi.stubEnv('VITE_APP_TWITCH_CLIENT_ID', 'VITE_APP_TWITCH_CLIENT_ID');
      vi.stubEnv('VITE_APP_TWITCH_CLIENT_SECRET', 'VITE_APP_TWITCH_CLIENT_SECRET');
      vi.stubEnv('VITE_APP_REDIRECT_URI', 'VITE_APP_REDIRECT_URI');
    });
    afterEach(()=>{
      vi.unstubAllEnvs();
    });
    test('Should render as expected', () => {
      const {container} = render(
        <Login />
      );
      expect(container).toMatchSnapshot();
    });
    test('Should render and handle modal events', async() => {
      vi.useFakeTimers({ toFake: ['nextTick'] });
      render(
        <Login />
      );
      const btnResetAppData = screen.getByText('Having issues?');
      const spanResetAppData = screen.getByText('Reset application data');
      expect(spanResetAppData.classList.contains('checkmark-display')).toBeFalsy();

      fireEvent.click(btnResetAppData);
      vi.advanceTimersByTime(1500);

      const dialogElement = await screen.findByRole('dialog');
      expect(dialogElement.classList.contains('show')).toBeTruthy();
      expect(dialogElement).toHaveTextContent('Are you sure you want to reset the application data');

      fireEvent.click(await screen.findByText('Cancel'));
      vi.advanceTimersByTime(1500);
      expect(dialogElement.classList.contains('show')).toBeFalsy();

      fireEvent.click(btnResetAppData);
      vi.advanceTimersByTime(1500);
      expect(dialogElement.classList.contains('show')).toBeTruthy();

      fireEvent.click(await screen.findByText('Reset'));
      vi.advanceTimersByTime(1500);
      expect(dialogElement.classList.contains('show')).toBeFalsy();

      expect(spanResetAppData.classList.contains('checkmark-display')).toBeTruthy();

      vi.useRealTimers();
    });
    test('Should render without error in production', () => {
      vi.stubEnv('DEV', false);
      vi.spyOn(window.localStorage, 'getItem').mockReturnValue('Error Message');
      const {container} = render(
        <Login />
      );
      expect(container).toMatchSnapshot();
    });
    test('Should render with error during development', () => {
      vi.stubEnv('DEV', true);
      vi.spyOn(window.localStorage, 'getItem').mockReturnValue('Error Message');
      const {container} = render(
        <Login />
      );
      expect(container).toMatchSnapshot();
    });
  });
});
