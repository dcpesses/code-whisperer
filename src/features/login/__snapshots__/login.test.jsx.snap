// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Login > render > Should render as expected 1`] = `
<div>
  <div
    class="container-fluid d-flex flex-column"
    id="login-screen"
  >
    <div
      class="row row-cols-auto align-items-center justify-content-center"
      id="landing"
    >
      <div
        class="col-8 col-md-6 text-center align-self-end align-self-md-center"
      >
        <div
          class="text-center pb-1"
        >
          <div>
            <img
              alt="Logo"
              class="login-logo w-50 mx-auto mb-3"
              src="/code-whisperer/src/assets/new-logo.svg"
            />
          </div>
          <h2
            class="text-center mb-0"
          >
            Code Whisperer
          </h2>
          <div
            class="text-center fs-6"
          >
            <small>
              Version 0.0.0
            </small>
          </div>
        </div>
      </div>
      <div
        class="col-9 col-md-6 text-center align-self-start align-self-md-center"
      >
        <a
          class="btn btn-sm fs-2 py-2 px-3 rounded-4 bg-gradient"
          href="https://id.twitch.tv/oauth2/authorize?client_id=VITE_APP_TWITCH_CLIENT_ID&response_type=code&scope=chat:read chat:edit moderation:read user:manage:whispers moderator:read:chatters user:read:moderated_channels channel:read:vips channel:read:editors moderator:manage:announcements user:read:subscriptions&redirect_uri=VITE_APP_REDIRECT_URI"
        >
          Log In With 
          <strong>
            Twitch
          </strong>
        </a>
        <div
          class="text-center fs-6"
        >
          <button
            class="btn btn-sm btn-link link-secondary text-decoration-none"
          >
            Having issues? 
            <span
              class="with-checkmark"
            >
              Reset application data
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Login > render > Should render with error during development 1`] = `
<div>
  <div
    class="container-fluid d-flex flex-column"
    id="login-screen"
  >
    <div
      class="row row-cols-auto align-items-center justify-content-center"
      id="landing"
    >
      <div
        class="col-8 col-md-6 text-center align-self-end align-self-md-center"
      >
        <div
          class="text-center pb-1"
        >
          <div>
            <img
              alt="Logo"
              class="login-logo w-50 mx-auto mb-3"
              src="/code-whisperer/src/assets/new-logo.svg"
            />
          </div>
          <h2
            class="text-center mb-0"
          >
            Code Whisperer
          </h2>
          <div
            class="text-center fs-6"
          >
            <small>
              Version 0.0.0
            </small>
          </div>
        </div>
      </div>
      <div
        class="col-9 col-md-6 text-center align-self-start align-self-md-center"
      >
        <div>
          <code
            class="small w-50 fs-6 mb-3"
          >
            Error: 
            Error Message
          </code>
        </div>
        <a
          class="btn btn-sm fs-2 py-2 px-3 rounded-4 bg-gradient"
          href="https://id.twitch.tv/oauth2/authorize?client_id=VITE_APP_TWITCH_CLIENT_ID&response_type=code&scope=chat:read chat:edit moderation:read user:manage:whispers moderator:read:chatters user:read:moderated_channels channel:read:vips channel:read:editors moderator:manage:announcements user:read:subscriptions&redirect_uri=VITE_APP_REDIRECT_URI"
        >
          Log In With 
          <strong>
            Twitch
          </strong>
        </a>
        <div
          class="text-center fs-6"
        >
          <button
            class="btn btn-sm btn-link link-secondary text-decoration-none"
          >
            Having issues? 
            <span
              class="with-checkmark"
            >
              Reset application data
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Login > render > Should render without error in production 1`] = `
<div>
  <div
    class="container-fluid d-flex flex-column"
    id="login-screen"
  >
    <div
      class="row row-cols-auto align-items-center justify-content-center"
      id="landing"
    >
      <div
        class="col-8 col-md-6 text-center align-self-end align-self-md-center"
      >
        <div
          class="text-center pb-1"
        >
          <div>
            <img
              alt="Logo"
              class="login-logo w-50 mx-auto mb-3"
              src="/code-whisperer/src/assets/new-logo.svg"
            />
          </div>
          <h2
            class="text-center mb-0"
          >
            Code Whisperer
          </h2>
          <div
            class="text-center fs-6"
          >
            <small>
              Version 0.0.0
            </small>
          </div>
        </div>
      </div>
      <div
        class="col-9 col-md-6 text-center align-self-start align-self-md-center"
      >
        <div>
          <code
            class="small w-50 fs-6 mb-3"
          >
            Error: 
            Error Message
          </code>
        </div>
        <a
          class="btn btn-sm fs-2 py-2 px-3 rounded-4 bg-gradient"
          href="https://id.twitch.tv/oauth2/authorize?client_id=VITE_APP_TWITCH_CLIENT_ID&response_type=code&scope=chat:read chat:edit moderation:read user:manage:whispers moderator:read:chatters user:read:moderated_channels channel:read:vips channel:read:editors moderator:manage:announcements user:read:subscriptions&redirect_uri=VITE_APP_REDIRECT_URI"
        >
          Log In With 
          <strong>
            Twitch
          </strong>
        </a>
        <div
          class="text-center fs-6"
        >
          <button
            class="btn btn-sm btn-link link-secondary text-decoration-none"
          >
            Having issues? 
            <span
              class="with-checkmark"
            >
              Reset application data
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
